sync/task_pool.go

```diff
@@ -158,15 +177,53 @@ func (p *TaskPool) run(id int, q chan task) error {
 	}
 }
 
-// add task
-func (p *TaskPool) AddTask(t task) {
-	id := atomic.AddUint32(&p.idx, 1) % uint32(p.tQNumber)
+// AddTask wait idle worker add task
+// return false when the pool is stop
+func (p *TaskPool) AddTask(t task) (ok bool) {
+	idx := atomic.AddUint32(&p.idx, 1)
+	id := idx % uint32(p.tQNumber)
 
 	select {
 	case <-p.done:
-		return
+		return false
+	default:
+		p.qArray[id] <- t
+		return true
+	}
+}
+
+// AddTaskAlways add task to queues or do it immediately
+func (p *TaskPool) AddTaskAlways(t task) {
+	id := atomic.AddUint32(&p.idx, 1) % uint32(p.tQNumber)
+
+	select {
 	case p.qArray[id] <- t:
+		return
+	default:
+		p.goSafely(t)
+	}
+}
+
+// AddTaskBalance add task to idle queue
+// do it immediately when no idle queue
+func (p *TaskPool) AddTaskBalance(t task) {
+	length := len(p.qArray)
+
+	// try len/2 times to lookup idle queue
+	for i := 0; i < length/2; i++ {
+		select {
+		case p.qArray[rand.Intn(length)] <- t:
```

Comment:
is it need to check p.done?

Reply:
method `AddTaskBalance` means `AddTaskAlwaysInBalance`